### PR TITLE
BUG: Fix the un-pickleable plot with DatetimeIndex

### LINKED ISSUE #18439: Can't pickle plots from dataframes with date indexes
Already report this at https://github.com/matplotlib/matplotlib/issues/9828, but it appears that that issue is caused by pandas using a lambda in the format_dateaxis function. I don't know if there is a way around this, but it prevents pandas-produced matplotlib figures from being pickled.

### MPL Bug report

Trying to pickle a figure that was produced using pandas.plot(), which includes a datetime index. Not sure if this is a pandas bug, or MPL.

Here's an example:
```
In [81]: sample
Out[81]: 
time
2002-07-28 08:00:00    0.0
2002-07-28 08:30:00    0.0
2002-07-28 09:00:00    0.0
2002-07-28 09:30:00    0.0
2002-07-28 10:00:00    0.0
2002-07-28 10:30:00    0.0
2002-07-28 11:00:00    0.0
2002-07-28 11:30:00    0.0
2002-07-28 12:00:00    0.0
2002-07-28 12:30:00    0.0
2002-07-28 13:00:00    0.0
2002-07-28 13:30:00    0.0
2002-07-28 14:00:00    0.0
2002-07-28 14:30:00    0.0
2002-07-28 15:00:00    0.0
2002-07-28 15:30:00    0.0
2002-07-28 16:00:00    NaN
2002-07-28 16:30:00    0.0
2002-07-28 17:00:00    0.0
2002-07-28 17:30:00    0.0
Name: Qle_qc, dtype: float64

In [82]: sample.plot()
Out[82]: <matplotlib.axes._subplots.AxesSubplot at 0x7fd5c06a1f60>

In [83]: fig = gcf()

In [84]: pickle.dump(fig, open('test.pkl', 'wb'))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/data/documents/uni/phd/projects/FluxnetTrafficLights/scripts/analysis/site_predictability.py in <module>()
----> 1 pickle.dump(fig, open('test.pkl', 'wb'))

AttributeError: Can't pickle local object 'format_dateaxis.<locals>.<lambda>'

In [85]: pickle.dump(sample, open('sample.pkl', 'wb'))

```

I've tried this with a manually constructed DF, and it worked, so I'm not sure why this is happening. I'm including a pickle of the above sample DF for investigation. It is originally a slice from an xarray, from a netcdf dataset.
[sample.pkl.gz](https://github.com/matplotlib/matplotlib/files/1493592/sample.pkl.gz)

**Matplotlib version**

Using Conda on linux (kubuntu 17.10):

```
ipython                   6.1.0                    py36_0  
matplotlib                2.1.0            py36hba5de38_0
pandas                    0.20.3                   py36_0  

MPL backend: Qt5Agg
```

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #18439
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
can u add a test as well

**Comment by Licht-T** [comment]
@jreback Where should I add test? `pandas/tests/plotting/*`?

**Comment by jreback** [comment]
sure

**Comment by jreback** [comment]
0.21.1 release note as well, yes ``pandas/tests/test_plotting`` is ok (or maybe in a misc section in plotting).

**Comment by Licht-T** [comment]
Thanks @jreback, fixed!

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by Licht-T** [comment]
CI failed because of the test. I'll check.

**Comment by Licht-T** [comment]
@jreback Seems that the pickling `instancemethod` only works in Python 3.

**Comment by Licht-T** [comment]
Now added the restriction to Python 3.

File: pandas/tests/plotting/test_datetimelike.py [hunk 1]
```diff
@@ -1470,5 +1471,10 @@ def _check_plot_works(f, freq=None, series=None, *args, **kwargs):
 
         with ensure_clean(return_filelike=True) as path:
             plt.savefig(path)
+
+        # GH18439
+        if PY3:
+            with ensure_clean(return_filelike=True) as path:
```
Comment (jreback) [inline pandas/tests/plotting/test_datetimelike.py]: could test for the round trip here, but prob overkill

File: pandas/tests/plotting/test_datetimelike.py [hunk 2]
```diff
@@ -1470,5 +1471,10 @@ def _check_plot_works(f, freq=None, series=None, *args, **kwargs):
 
         with ensure_clean(return_filelike=True) as path:
             plt.savefig(path)
+
+        # GH18439
```
Comment (jreback) [inline pandas/tests/plotting/test_datetimelike.py]: why is this only PY3?

add a more informative comment, not just the issue, e.g. ensure round-trip pickle compat

Comment (jreback) [reply pandas/tests/plotting/test_datetimelike.py]: @Licht-T saw your comment, ok then, just make a comment that we are pickling instancemethods which is supported on PY3 only (hmm thought this did work on PY2)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you update

**Comment by Licht-T** [comment]
@jreback Okay! I'll re-check the pickling issue on Python 2 and update until tomorrow.

**Comment by jreback** [comment]
can you update

**Comment by Licht-T** [comment]
Sorry @jreback! Now updated!

**Comment by jreback** [comment]
lgtm
ping on green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**